### PR TITLE
Allow editing scheduled emails

### PR DESCRIPTION
**Pull request by tomassatinsky**
Jira: TNT-516

---

Supported PR commands:

| Command                  | Description             |
| ------------------------ | ----------------------- |
| `ok to test`             | Re-run standard checks  |
| `extended test`          | BackstopJS tests        |
| `extended check sonar`   | SonarQube tests         |
| `extended check cypress` | Cypress E2E tests       |
| `extended check plugins` | Dashboard plugins tests |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/7503/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9364/ : FAILURE in 16m 28s

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/utils/repeat.ts [hunk 1]
```diff
@@ -151,3 +150,82 @@ export function setWeeklyRepeat(repeatData: IScheduleEmailRepeat, startDate: Dat
...
+function getFirstFragment(fragments: number[]): { firstFragmentIndex: number; firstFragmentValue: number } {
+    for (let index = 0; index < fragments.length; index++) {
+        const fragment = fragments[index];
+        if (fragment !== 0) {
+            return {
+                firstFragmentIndex: index,
+                firstFragmentValue: fragment || 0,
+            };
+        }
+    }
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/utils/repeat.ts:170]: How about using `findIndex` and returning it together with `fragments[index]`?

File: libs/sdk-ui-dashboard/src/model/commandHandlers/scheduledEmail/saveScheduledEmailHandler.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// (C) 2021 GoodData Corporation
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/scheduledEmail/saveScheduledEmailHandler.ts:1]: Keep it up to date please

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts [hunk 3]
```diff
@@ -111,14 +117,19 @@ interface UseScheduledEmailResult {
...
+     * Status of the scheduled email creation
      */
     scheduledEmailCreationStatus?: CommandProcessingStatus;
 
     /**
-     * Attachment to be selected by default.
+     * Function that results in saving the existing scheduled email on the backend.
      */
-    defaultAttachment?: ObjRef;
+    handleSaveScheduledEmail: (scheduledEmailToCreate: IScheduledMailDefinition) => void;
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts:127]: I would rename the `scheduledEmailToCreate` to clearly represent already created schedule. Maybe `scheduledEmail` is enough.

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts [hunk 4]
```diff
@@ -187,9 +213,22 @@ export const useScheduledEmail = (props: UseScheduledEmailProps): UseScheduledEm
...
+        onSuccess: onSaveSuccess,
+        onError: onSaveError,
+        onBeforeRun: onSave,
+    });
+
+    const handleSaveScheduledEmail = useCallback(
+        (scheduledEmail: IScheduledMailDefinition) => {
+            scheduledEmailSaver.save(scheduledEmail);
+        },
+        [filters, hasDefaultFilters],
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts:228]: Are the dependecies here on purpose? I don't see any reason for them.

File: libs/sdk-ui-dashboard/src/model/react/useDashboardScheduledEmails.ts [hunk 5]
```diff
@@ -46,6 +46,8 @@ export const useDashboardScheduledEmails = () => {
         dispatch(uiActions.setScheduleEmailDialogDefaultAttachment(attachmentRef));
     const resetScheduledEmailDefaultAttachment = () =>
         enableInsightExportScheduling && dispatch(uiActions.resetScheduleEmailDialogDefaultAttachment());
+    const [emailingManagementEditedSchedule, setEmailingManagementEditedSchedule] =
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/model/react/useDashboardScheduledEmails.ts:49]: I would rename this to something like `scheduledEmailToEdit`

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/utils/repeat.ts (hunk 1 above)
Comment (tomassatinsky) [reply libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/utils/repeat.ts]: 👍 done

File: libs/sdk-ui-dashboard/src/model/commandHandlers/scheduledEmail/saveScheduledEmailHandler.ts (hunk 2 above)
Comment (tomassatinsky) [reply libs/sdk-ui-dashboard/src/model/commandHandlers/scheduledEmail/saveScheduledEmailHandler.ts]: done

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts (hunk 3 above)
Comment (tomassatinsky) [reply libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts]: yes, my mistake, fixed

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/utils/repeat.ts [hunk 6]
```diff
@@ -158,17 +159,12 @@ const HOUR_INDEX = 4;
...
-function getFirstFragment(fragments: number[]): { firstFragmentIndex: number; firstFragmentValue: number } {
-    for (let index = 0; index < fragments.length; index++) {
-        const fragment = fragments[index];
-        if (fragment !== 0) {
-            return {
-                firstFragmentIndex: index,
-                firstFragmentValue: fragment || 0,
-            };
-        }
+function getFirstFragment(fragments: number[]): number {
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/utils/repeat.ts:162]: Maybe also rename this to `getFirstFragmentIndex` now

**Review by scavnickyj — APPROVED** [review]

**Comment by scavnickyj** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/7520/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9382/ : SUCCESS in 17m 55s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2419

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2177/ : SUCCESS in 57s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2463/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1287/ : SUCCESS in 22m 14s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9386/ : SUCCESS in 16m 03s

**Pull request merged**